text/0000-target-bundles.md

```diff
@@ -0,0 +1,245 @@
+    "no-default-libraries": true,
+    "custom-unwind-resume": false,                          // might make sense to merge into below
+    "eh-method": "dwarf",                                   // NEW: what EH method to use
+    "dll-storage-attrs": false,                             // NEW: should use dll storage attrs?
+    "debug-info": ["Dwarf", 4],                             // NEW: what debug info format
+    "system-abi": "C",                                      // NEW: what "system" ABI means
+    "c-abi-kind": "cabi_x86_64",                            // NEW: what C-ABI implementation to use
+
+    // Various stuff adjusting how linkage is done
+    "pre-link-args": ["-Wl,--as-needed", "-Wl,-z,noexecstack"],
+    "post-link-args": [],
+    "pre-link-objects-dll": [],
+    "pre-link-objects-exe": [],
+    "post-link-objects": [],
+    "late-link-args": [],
+    "gc-sections-args": [],                                 // NEW: how to strip sections
+    "rpath-prefix": "$ORIGIN",                              // CHANGED: to allow specifying rpath prefix, null to disable rpath altogether
+    "no-compiler-rt": false,
+    "metadata-section": ".note.rustc",                      // NEW: name of section for metadata
+    "has-frameworks": false,                                // NEW: is concept of frameworks supported?
+    "position-independent-executables": true,               // should become a plain linker argument?
+    "lib_allocation_crate": "alloc_system",
+    "exe_allocation_crate": "alloc_jemalloc",
+    // should become templates? `lib{}.so` is much nicer
+    "dll_prefix": "lib",
+    "dll_suffix": ".so",
+    "exe_suffix": "",
+    "staticlib_prefix": "lib",
+    "staticlib_suffix": ".a",
+
+    // LLVM options
+    "cpu": "x86-64",                                        // CPU of the target
+    "features": "",                                         // LLVM features
+    "relocation_model": "pic",
+    "code_model": "default",
+    "eliminate-frame-pointer": true,
+
+    // Apparently OS X is so much of a special case that even moving all the special cases into
+    // configuration is painful.
+    is_like_osx: false,
+
+    // is_like_solaris is handled by the extra gc-sections key
+    // is_like_msvc is handled by the extra metadata-section, linker-kind, eh-method,
+    // dll-storage-attrs, debug-info keys
+    // is_like_windows is handled by the extra system-abi and c-abi-kind keys
+    // is_like_android is handled by the extra debug-info key
+}
+```
+
+## Distribution of targets
+
+Currently every built-in target is distributed along with the rustc compiler. This is suboptimal,
+because in majority of cases users are interested in targets which rustc can compile for, rather
+than the built-in targets rustc knows about, therefore distributing targets built-in into rustc is
+providing no benefits.
+
+Native libraries and linkers aside, it is obvious there’s little sense in distributing targets
```

Comment:
I don't agree with this. Once "on the fly compilation of std" becomes a thing, it actually makes sense to provide all the target specifications "upfront" with `rustc`. Then people, specially the ones working with/on embedded systems, can choose which std features to enable/disable (e.g. remove RUST_BACKTRACE support) (\*) and/or compile std with CPU specific optimizations for maximum performance (\*\*). In this scenario, rust-std becomes a convenience to avoid the compile time required to build std the first time.

(\*) This actually makes more sense for final deployments/releases.
(\*\*) This makes sense for both development (slightly faster execution of test suites but the savings add up quickly) and for deployment to embedded systems. For releases that must work everywhere, one should instead not enable any CPU-specific optimization -- this would be the default, zero-configuration option, which is also how rust-std is compiled.

Instead of what's proposed here, I propose adding a new `targets` component that will be installed by `rustup` *once* regardless of which or how many toolchains (`rustc` components) one installs. After all, target specifications are indepedent of the host system. This also means one less step (`rustup target add foo`) in the initial cross compilation setup (once "on the fly compilation of std" is here).